Applications/Spire/Source/Ui/ColorPicker.cpp

```diff
@@ -0,0 +1,336 @@
+      make_modifiers(), make_modifiers());
+    color_spectrum->setFixedHeight(scale_height(164));
+    color_spectrum->findChild<QLabel*>()->setScaledContents(false);
+    update_style(*color_spectrum, [&] (auto& style) {
+      style.get(Any()).set(border(scale_width(1), QColor(0xC8C8C8)));
+      style.get(Any() > Thumb() > is_a<Icon>()).
+        set(Fill(optional<QColor>())).
+        set(IconImage(get_thumb_icon()));
+      style.get(FocusVisible() > Thumb() > is_a<Icon>()).
+        set(IconImage(get_thumb_invert_icon()));
+    });
+    return color_spectrum;
+  }
+
+  auto make_hue_slider(std::shared_ptr<DecimalModel> model) {
+    auto hue_slider = new Slider(std::move(model), make_modifiers());
+    hue_slider->setFixedHeight(scale_height(16));
+    update_style(*hue_slider, [&] (auto& style) {
+      style.get(Any()).set(border(scale_width(1), QColor(0xC8C8C8)));
+      style.get(Any() > Track()).set(IconImage(get_hue_spectrum_image()));
+      style.get(Any() > Thumb() > is_a<Icon>()).
+        set(Fill(optional<QColor>())).
+        set(IconImage(get_thumb_icon()));
+      style.get(FocusVisible() > Thumb() > is_a<Icon>()).
+        set(IconImage(get_thumb_invert_icon()));
+    });
+    return hue_slider;
+  }
+
+  auto make_alpha_slider(std::shared_ptr<DecimalModel> model) {
+    auto alpha_slider = new Slider(std::move(model), make_modifiers());
+    alpha_slider->setFixedHeight(scale_height(16));
+    alpha_slider->findChild<QLabel*>()->setScaledContents(false);
+    match(*alpha_slider, Alpha());
+    update_style(*alpha_slider, [&] (auto& style) {
+      style.get(Any()).set(border(scale_width(1), QColor(0xC8C8C8)));
+      style.get(Any() > Thumb() > is_a<Icon>()).
+        set(Fill(boost::optional<QColor>())).
+        set(IconImage(get_thumb_icon()));
+      style.get(FocusVisible() > Thumb() > is_a<Icon>()).
+        set(IconImage(get_thumb_invert_icon()));
+    });
+    return alpha_slider;
+  }
+}
+
+struct ColorPicker::ColorPickerModel {
+  std::shared_ptr<ValueModel<QColor>> m_source;
+  std::shared_ptr<LocalDecimalModel> m_hue_slider_model;
+  std::shared_ptr<LocalDecimalModel> m_alpha_slider_model;
+  std::shared_ptr<LocalDecimalModel> m_spectrum_x_model;
+  std::shared_ptr<LocalDecimalModel> m_spectrum_y_model;
+  scoped_connection m_source_connection;
+  scoped_connection m_hue_connection;
+  scoped_connection m_alpha_connection;
+  scoped_connection m_spectrum_x_connection;
+  scoped_connection m_spectrum_y_connection;
+
+  ColorPickerModel(std::shared_ptr<ValueModel<QColor>> source)
+      : m_source(std::move(source)),
+        m_hue_slider_model(std::make_shared<LocalDecimalModel>()),
+        m_alpha_slider_model(std::make_shared<LocalDecimalModel>()),
+        m_spectrum_x_model(std::make_shared<LocalDecimalModel>()),
+        m_spectrum_y_model(std::make_shared<LocalDecimalModel>()) {
+    m_hue_slider_model->set_minimum(Decimal(0));
```

Comment:
Does it fail if you omit the ```Decimal```?

    m_hue_slider_model->set_minimum(0);
    m_hue_slider_model->set_maximum(360);

Reply:
Yes, because `set_minimum` and `set_maximum` of those models accept `boost::optional` as argument.